src/Storages/StorageMergeTree.cpp

```diff
@@ -2079,47 +2082,96 @@ void StorageMergeTree::replacePartitionFrom(const StoragePtr & source_table, con
+
+    for (DataPartPtr & src_part : src_parts)
     {
         if (!canReplacePartition(src_part))
             throw Exception(ErrorCodes::BAD_ARGUMENTS,
                             "Cannot replace partition '{}' because part '{}' has inconsistent granularity with table",
-                            partition_id, src_part->name);
+                            source_partition_id, src_part->name);
 
+        IDataPartStorage::ClonePartParams clone_params{.txn = local_context->getCurrentTransaction()};
         /// This will generate unique name in scope of current server process.
-        Int64 temp_index = insert_increment.get();
-        MergeTreePartInfo dst_part_info(partition_id, temp_index, temp_index, src_part->info.level);
+        auto index = insert_increment.get();
 
-        IDataPartStorage::ClonePartParams clone_params{.txn = local_context->getCurrentTransaction()};
-        auto [dst_part, part_lock] = cloneAndLoadDataPartOnSameDisk(
-            src_part,
-            TMP_PREFIX,
-            dst_part_info,
-            my_metadata_snapshot,
-            clone_params,
-            local_context->getReadSettings(),
-            local_context->getWriteSettings());
-        dst_parts.emplace_back(std::move(dst_part));
-        dst_parts_locks.emplace_back(std::move(part_lock));
-    }
-
-    /// ATTACH empty part set
-    if (!replace && dst_parts.empty())
-        return;
+        MergeTreePartInfo dst_part_info(destination_partition_id, index, index, src_part->info.level);
+
+        if (is_partition_exp_the_same)
+        {
+            auto [dst_part, part_lock] = cloneAndLoadDataPartOnSameDisk(
+                src_part,
+                TMP_PREFIX,
+                dst_part_info,
+                my_metadata_snapshot,
+                clone_params,
+                local_context->getReadSettings(),
+                local_context->getWriteSettings());
+            dst_parts.emplace_back(std::move(dst_part));
+            dst_parts_locks.emplace_back(std::move(part_lock));
+        }
+        else
+        {
+            auto metadata_manager = std::make_shared<PartMetadataManagerOrdinary>(src_part.get());
+            IMergeTreeDataPart::MinMaxIndex destination_min_max_index;
+
+            destination_min_max_index.load(src_data, metadata_manager);
+
+            auto [dst_part, part_lock] = cloneAndLoadPartOnSameDiskWithDifferentPartitionKey(
+                src_part,
+                destination_partition,
+                dst_part_info,
+                destination_min_max_index,
+                TMP_PREFIX,
+                my_metadata_snapshot,
+                clone_params,
+                local_context);
+
+            dst_parts.emplace_back(std::move(dst_part));
+            dst_parts_locks.emplace_back(std::move(part_lock));
+        }
+    }
 
     MergeTreePartInfo drop_range;
     if (replace)
     {
-        drop_range.partition_id = partition_id;
+        drop_range.partition_id = destination_partition_id;
```

Comment:
This does not make a difference. It won't reach this code if `replace == true` and `is_partition_expression_the_same == false`